### ISSUE TITLE
Hide replies to comments that have been spammed or aren't yet moderated

### ISSUE THREAD

**Issue by jywarren** [publiclab/plots2#8854]
It looks like replies to comments are not filtered for spam. We should see this message: 

“Are you sure? The user will no longer be able to log in or publish, and their content will be hidden except comments.” seems wrong

Example of this happening: https://publiclab.org/notes/bhamster/09-02-2020/public-lab-virtual-event-on-all-things-microplastics 

See this reply to a comment:

![image](https://user-images.githubusercontent.com/24359/102243195-e3227680-3ec8-11eb-94c2-165ac86514a3.png)

Instead it should have a message like this, which is working for regular comments (those that reply to a post, not another comment:

![image](https://user-images.githubusercontent.com/24359/102243313-05b48f80-3ec9-11eb-919a-d2dd39d84c45.png)

### ISSUE
(This might be fixed by now...lets write a test to confirm that and fix any failures that may arise from the test):

Replies to comments aren’t filtering out spam comments. We use this `replied_comments` method in several places, so we may need to make this fix several times. 

https://github.com/publiclab/plots2/search?q=replied_comments

We can't do it in the definition because it's actually an ActiveRecord relation, not a custom method we can add filters to: https://github.com/publiclab/plots2/blob/828321c2a8fd91e7a55140a511ed4e6dda718b61/app/models/comment.rb#L10

We should filter out `status = 0` in the definition of `comment.replied_comments`, and also for `status = 4` show the message above. 

Then, we ought to write some basic tests: 

https://github.com/publiclab/plots2/blob/876d0fc084064aaecc23f8003630d7d1ab858fa1/test/functional/comment_controller_test.rb#L57-L61

**Comment by noi5e** [comment]
This seems somewhat in my wheelhouse! I can break it down into sub-issues at a moment when I feel ahead of my work.

**Comment by cesswairimu** [comment]
I was testing this locally to break it to a ftos, it looks like the issue is already fixed. Could someone also confirm the same is the case on their end... @noi5e  maybe you solved it somewhere in your project..

**Comment by noi5e** [comment]
@cesswairimu Hmm!! I think someone else did 😅

**Comment by noi5e** [comment]
@cesswairimu Someone could write a test for this, if I don't get to it? That's a little too complicated for an FTO, but I think writing tests can teach applicants a lot. Also maybe we shouldn't close this issue until a test is written

**Comment by cesswairimu** [comment]
@noi5e I agree, I will update this to be more of a test-writing issue. Thanks

**Comment by jywarren** [comment]
Hmm, i'm not sure -- maybe the test can come first to confirm if it's really fixed? 

Because we can only filter upon usage, not in the `has_many` definition, it seems, and I don't see a `.where.not(status: 0)` on these:

https://github.com/publiclab/plots2/blob/ebf57e0cdd1dcc9d5cc884a15e7cd5b5bbe78390/app/views/notes/_comments.html.erb#L8

https://github.com/publiclab/plots2/blob/401874455490c6300270ef54490fe9452e0ca522/app/views/questions/show.html.erb#L69

https://github.com/publiclab/plots2/blob/ebf57e0cdd1dcc9d5cc884a15e7cd5b5bbe78390/app/views/notes/_comment.html.erb#L162

What do you think?

Thanks, all!!! 💯

**Comment by jywarren** [comment]
ah and i guess on those `includes()` ones, it would have to be named, so maybe... `.where('comments.status != 0')` ?

Yes i think that's correct, see usage here:

https://github.com/publiclab/plots2/blob/672d63ab11c2159e28d2cc502c47a3b3c83730b9/app/models/node.rb#L1114

**Comment by 17sushmita** [comment]
May I work on writing tests for this?

**Comment by noi5e** [comment]
@17sushmita Yes, definitely, please go ahead, that would be much appreciated!

**Comment by 17sushmita** [comment]
> It looks like replies to comments are not filtered for spam. We should see this message:
> 
> “Are you sure? The user will no longer be able to log in or publish, and their content will be hidden except comments.” seems wrong
@jywarren What does this mean by saying this message seems wrong? Is it not supposed to be displayed when moderator tries to mark the comment as spam? Kindly give a little more context to it.

**Comment by noi5e** [comment]
@17sushmita I'm a little confused by that too. I think maybe Jeffrey meant that text was appearing in yellow in the image below, instead of `Moderate first-time comment`? If so, that does seem wrong.

![102243313-05b48f80-3ec9-11eb-919a-d2dd39d84c45](https://user-images.githubusercontent.com/4361605/111917115-75dc2f80-8a3b-11eb-9c0b-a0548bd47f0a.png)

While we wait for clarification, you can still go ahead and write a test. I'm just going to post some suggestions here that might be helpful.

To start with, you can insert a spam/first-time-poster comment into the testing database that is a **reply** to another comment:
```
nodes(node_name).add_comment({
  uid: user_id,
  body: comment_text
  reply_to: parent_comment_id
  status: 0 # comment status: 0 for banned, 4 for first-time poster (a moderator has to approve the comment)
})
```

Then write the rest of the test from there, making sure that the method we use to retrieve comments isn't getting comment replies with status 0 or 4.

Something like that! You'll probably have to do some research. Definitely leave a comment here if you get stuck, or need some pointers for where to look in the codebase. We're here to help!

**Comment by 17sushmita** [comment]
@noi5e , Thanks a lot for helping out 😃️!! I need one more clarification. Do I need to create a new file for comment reply tests or make changes in maybe [/plots2/test/functional/admin_controller_test.rb](https://github.com/publiclab/plots2/blob/main/test/functional/admin_controller_test.rb) or [/plots2/test/system/spam2_test.rb](https://github.com/publiclab/plots2/blob/main/test/system/spam2_test.rb) or [/plots2/test/functional/comment_controller_test.rb](https://github.com/publiclab/plots2/blob/main/test/functional/comment_controller_test.rb) or any other file?

**Comment by noi5e** [comment]
@17sushmita I'm not 100% sure on this, but I'm thinking `/test/functional/comment_controller_test.rb`. The other place I was thinking was a unit test for comments. Thoughts from anyone else?

In the meantime, feel free to start working on this while we wait for others to weigh in.

**Comment by 17sushmita** [comment]
> @17sushmita I'm not 100% sure on this, but I'm thinking `/test/functional/comment_controller_test.rb`. The other place I was thinking was a unit test for comments. Thoughts from anyone else?
> 
> In the meantime, feel free to start working on this while we wait for others to weigh in.

Thanks, I started writing the tests, but one thing I observed is that even after marking the comment as spam, it does show up and according to the software it is correct because while marking it spam, it generates a warning as mentioned by @jywarren above "Are you sure? The user will no longer be able to log in or publish, and their content will be hidden **except comments**." but, is it really appropriate to display spam marked comments?🤔️

**Comment by noi5e** [comment]
> is it really appropriate to display spam marked comments?🤔️

I don't think so! 

This text you mention is pretty interesting:

> "Are you sure? The user will no longer be able to log in or publish, and their content will be hidden **except comments.**" 

It makes me think that maybe back in the day, we didn't moderate comments. I think that piece of text is just outdated.

Whatever the case I think we all agree that we definitely should filter all comments! So we can avoid the spambots that @jywarren [originally mentioned](https://github.com/publiclab/plots2/issues/8854#issue-767783460).

I just tested this locally, and to me it looks like this hasn't been fixed. We currently aren't filtering spam or first-time moderated comment **REPLIES**... Please keep in mind that regular comments (that are not **REPLIES** to other comments) are being filtered, so this just applies to replies. (I tested for comments that are NOT replies, and those are being moderated)

Going to break this down a little further. Locally, I created a fresh account, and posted a comment on a note. This is what I see when I post a comment. So far so good:

<img width="610" alt="Screen Shot 2021-03-22 at 11 39 30 AM" src="https://user-images.githubusercontent.com/4361605/112045091-b0ab9980-8b07-11eb-8ffb-f95f590fd966.png">

As a user, the _"Pending approval by community moderators"_ message makes me think that my comment won't appear publicly.

However, the comment **DOES** appear publicly! When I visit the same note in an incognito window that's logged out of the site, the comment appears exactly as it appeared to user _cisco_:

<img width="485" alt="Screen Shot 2021-03-22 at 11 43 21 AM" src="https://user-images.githubusercontent.com/4361605/112045521-2dd70e80-8b08-11eb-8d5b-e43f6bf07316.png">

Banning user _cisco_ doesn't seem to change things much either. The _"Pending approval"_ message disappears, but the comment is still viewable from an incognito window:

<img width="543" alt="Screen Shot 2021-03-22 at 11 46 06 AM" src="https://user-images.githubusercontent.com/4361605/112045625-49421980-8b08-11eb-961e-033b30a3d23c.png">

So yeah, this definitely needs to change.

@17sushmita Go ahead and keep working on this! 

Also, if you haven't found it yet, we have a banned _spammer_ user already created in `/test/fixtures/users.yml`

**Comment by 17sushmita** [comment]
@noi5e That was a very clear explination. I too tested very similar on my local system and results were same which made me think so. So, I'll first work to remove the spam comments and then write test for it if it is fixed or not.

**Comment by noi5e** [comment]
As an aside, it looks like comment replies don't currently appear in admin's spam moderation dashboard:

<img width="744" alt="Screen Shot 2021-03-22 at 11 40 21 AM" src="https://user-images.githubusercontent.com/4361605/112048229-6cba9380-8b0b-11eb-940e-0b3c82adf20a.png">

And if user cisco posts a comment reply, it doesn't make them show up in Active Users:

<img width="458" alt="Screen Shot 2021-03-22 at 11 40 35 AM" src="https://user-images.githubusercontent.com/4361605/112048299-7d6b0980-8b0b-11eb-89cf-0e9f4755bb51.png">

Potential issue for the Outreachy spam moderation project mentioned in #9257?

**Comment by jywarren** [comment]
Thank you @17sushmita and @noi5e for so thoroughly unpacking this. My apologies for the unclear note in my original message but @noi5e is completely correct - we just need the simpler "Moderate first-time comment" message. 

> filtering spam or first-time moderated comment REPLIES

And great catch that still is unsolved for REPLIES.

Super appreciate all your help, both of you. 🎉 

I'll note the spam dashboard issue in the new project description!!!

**Comment by 17sushmita** [comment]
Hi @jywarren @noi5e, Sorry for asking again but I had one more confusion. Should we display a comment which is not spam but the user is banned? What to do in such cases?

**Comment by jywarren** [comment]
Yes, currently the status of banned users doesnt affect visibility of their
comments. Thank you!

**Comment by TildaDares** [comment]
> As an aside, it looks like comment replies don't currently appear in admin's spam moderation dashboard:
> 
> <img alt="Screen Shot 2021-03-22 at 11 40 21 AM" width="744" src="https://user-images.githubusercontent.com/4361605/112048229-6cba9380-8b0b-11eb-940e-0b3c82adf20a.png">
> 
> And if user cisco posts a comment reply, it doesn't make them show up in Active Users:
> 
> <img alt="Screen Shot 2021-03-22 at 11 40 35 AM" width="458" src="https://user-images.githubusercontent.com/4361605/112048299-7d6b0980-8b0b-11eb-89cf-0e9f4755bb51.png">
> 
> Potential issue for the Outreachy spam moderation project mentioned in #9257?

@noi5e I can't reproduce the comment replies issue locally. I logged in as a new user, made a comment reply, logged out and then logged in as admin and then I marked the comment reply as spam. I went to the spam moderation page and I found the comment reply under comments.

![image](https://user-images.githubusercontent.com/63427719/115221451-99110200-a101-11eb-940e-cbc5e32a9f0e.png)

![image](https://user-images.githubusercontent.com/63427719/115221483-a0d0a680-a101-11eb-8e35-63eab542c462.png)

Is there something I'm missing?

**Comment by jcads** [comment]
Hi @TildaDares, i think this issue has been fixed in #9381

**Comment by TildaDares** [comment]
Thanks @jcads
